File: src/kitchen_sync/sync.py
"""Core sync engine: staging, symlink resolution, and tree diffing."""

from __future__ import annotations

import difflib
import os
import shutil
import subprocess
import tempfile
from pathlib import Path
from typing import Any


def prepare_staging(
    source_dir: Path,
    sync_paths: list[str],
    resolve_symlinks: bool = False,
) -> Path:
    """Copy sync_paths from source_dir into a temp staging directory.

    If resolve_symlinks is True, symlinks are followed and actual content is
    copied. Otherwise symlinks are preserved as-is.
    """
    staging = Path(tempfile.mkdtemp(prefix="ksync-"))

    for sp in sync_paths:
        src = source_dir / sp
        if not src.exists():
            continue

        dst = staging / sp

        if src.is_file() or (src.is_symlink() and not src.is_dir()):
            dst.parent.mkdir(parents=True, exist_ok=True)
            if resolve_symlinks and src.is_symlink():
                # Copy the resolved content
                real = src.resolve()
                if real.is_dir():
                    shutil.copytree(real, dst)
                else:
                    shutil.copy2(real, dst)
            elif src.is_symlink() and not resolve_symlinks:
                os.symlink(os.readlink(src), dst)
            else:
                shutil.copy2(src, dst)
        elif src.is_dir():
            _copy_dir(src, dst, resolve_symlinks)

    return staging


def _copy_dir(src: Path, dst: Path, resolve_symlinks: bool) -> None:
    """Recursively copy a directory, optionally resolving symlinks."""
    dst.mkdir(parents=True, exist_ok=True)

    for item in src.iterdir():
        item_dst = dst / item.name

        if item.is_symlink():
            if resolve_symlinks:
                real = item.resolve()
                if real.is_dir():
                    shutil.copytree(real, item_dst)
                elif real.is_file():
                    shutil.copy2(real, item_dst)
                # Skip broken symlinks
            else:
                os.symlink(os.readlink(item), item_dst)
        elif item.is_dir():
            _copy_dir(item, item_dst, resolve_symlinks)
        elif item.is_file():
            shutil.copy2(item, item_dst)


def collect_files(base_dir: Path, sync_paths: list[str]) -> dict[str, str]:
    """Collect all files under sync_paths relative to base_dir.

    Returns {relative_path: content} for text files.
    """
    files = {}
    for sp in sync_paths:
        src = base_dir / sp
        if not src.exists():
            continue
        if src.is_file():
            try:
                files[sp] = src.read_text()
            except (UnicodeDecodeError, PermissionError):
                files[sp] = "<binary>"
        elif src.is_dir():
            for f in sorted(src.rglob("*")):
                if f.is_file():
                    rel = str(f.relative_to(base_dir))
                    try:
                        files[rel] = f.read_text()
                    except (UnicodeDecodeError, PermissionError):
                        files[rel] = "<binary>"
    return files


def diff_trees(
    local_files: dict[str, str],
    remote_files: dict[str, str],
) -> list[str]:
    """Generate unified diffs between local and remote file trees.

    Returns a list of diff strings (one per changed file).
    """
    all_paths = sorted(set(local_files) | set(remote_files))
    diffs = []

    for path in all_paths:
        local_content = local_files.get(path, "")
        remote_content = remote_files.get(path, "")

        if local_content == remote_content:
            continue

        if path not in remote_files:
            label = f"  + {path} (local only)"
            diffs.append(label)
            continue

        if path not in local_files:
            label = f"  - {path} (remote only)"
            diffs.append(label)
            continue

        diff_lines = difflib.unified_diff(
            remote_content.splitlines(keepends=True),
            local_content.splitlines(keepends=True),
            fromfile=f"remote/{path}",
            tofile=f"local/{path}",
        )
        diff_text = "".join(diff_lines)
        if diff_text:
            diffs.append(diff_text)

    return diffs


def git_status_for_paths(
    repo_dir: Path,
    sync_paths: list[str],
    git_env: dict[str, str] | None = None,
) -> dict[str, Any]:
    """Get git status information scoped to sync_paths.

    Returns dict with:
      - modified: list of modified/untracked files
      - ahead: number of commits ahead of remote
      - behind: number of commits behind remote
    """
    env = {**os.environ, **(git_env or {})}

    # Get status for sync paths
    cmd = ["git", "status", "--porcelain", "--"] + sync_paths
    result = subprocess.run(
        cmd, capture_output=True, text=True, cwd=repo_dir, env=env
    )
    modified = []
    for line in result.stdout.strip().splitlines():
        if line:
            modified.append(line.strip())

    # Get ahead/behind
    ahead = 0
    behind = 0
    result = subprocess.run(
        ["git", "rev-list", "--left-right", "--count", "HEAD...@{upstream}"],
        capture_output=True,
        text=True,
        cwd=repo_dir,
        env=env,
    )
    if result.returncode == 0 and result.stdout.strip():
        parts = result.stdout.strip().split()
        if len(parts) == 2:
            ahead = int(parts[0])
            behind = int(parts[1])

    return {"modified": modified, "ahead": ahead, "behind": behind}


def git_push(
    repo_dir: Path,
    sync_paths: list[str],
    message: str,
    git_env: dict[str, str] | None = None,
    dry_run: bool = False,
) -> dict[str, Any]:
    """Stage sync_paths, commit, and push.

    Returns dict with commit_hash and files_staged.
    """
    env = {**os.environ, **(git_env or {})}

    # Stage files
    cmd = ["git", "add", "--"] + sync_paths
    subprocess.run(cmd, cwd=repo_dir, env=env, check=True)

    # Check if there's anything to commit
    result = subprocess.run(
        ["git", "diff", "--cached", "--name-only"],
        capture_output=True,
        text=True,
        cwd=repo_dir,
        env=env,
    )
    staged = [f for f in result.stdout.strip().splitlines() if f]

    if not staged:
        return {"commit_hash": None, "files_staged": []}

    if dry_run:
        # Unstage
        subprocess.run(
            ["git", "reset", "HEAD", "--"] + sync_paths,
            cwd=repo_dir,
            env=env,
            capture_output=True,
        )
        return {"commit_hash": "(dry-run)", "files_staged": staged}

    # Commit
    subprocess.run(
        ["git", "commit", "-m", message],
        cwd=repo_dir,
        env=env,
        check=True,
        capture_output=True,
    )

    # Get commit hash
    result = subprocess.run(
        ["git", "rev-parse", "--short", "HEAD"],
        capture_output=True,
        text=True,
        cwd=repo_dir,
        env=env,
    )
    commit_hash = result.stdout.strip()

    # Push
    subprocess.run(
        ["git", "push"],
        cwd=repo_dir,
        env=env,
        check=True,
        capture_output=True,
    )

    return {"commit_hash": commit_hash, "files_staged": staged}


def git_pull(
    repo_dir: Path,
    git_env: dict[str, str] | None = None,
    dry_run: bool = False,
) -> str:
    """Pull latest from remote. Returns output message."""
    env = {**os.environ, **(git_env or {})}

    if dry_run:
        result = subprocess.run(
            ["git", "fetch", "--dry-run"],
            capture_output=True,
            text=True,
            cwd=repo_dir,
            env=env,
        )
        return result.stderr.strip() or "Already up to date."

    result = subprocess.run(
        ["git", "pull"],
        capture_output=True,
        text=True,
        cwd=repo_dir,
        env=env,
        check=True,
    )
    return result.stdout.strip() or "Already up to date."

File: src/sync.ts
import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import { execSync, spawnSync } from "child_process";
import { createTwoFilesPatch } from "diff";

export function prepareStaging(
  sourceDir: string,
  syncPaths: string[],
  resolveSymlinks = false,
): string {
  const staging = fs.mkdtempSync(path.join(os.tmpdir(), "ksync-"));

  for (const sp of syncPaths) {
    const src = path.join(sourceDir, sp);
    if (!fs.existsSync(src)) continue;

    const dst = path.join(staging, sp);
    const stat = fs.lstatSync(src);

    if (stat.isFile() || (stat.isSymbolicLink() && !fs.statSync(src).isDirectory())) {
      fs.mkdirSync(path.dirname(dst), { recursive: true });
      if (resolveSymlinks && stat.isSymbolicLink()) {
        const real = fs.realpathSync(src);
        const realStat = fs.statSync(real);
        if (realStat.isDirectory()) {
          copyDirRecursive(real, dst, false);
        } else {
          fs.copyFileSync(real, dst);
        }
      } else if (stat.isSymbolicLink() && !resolveSymlinks) {
        fs.symlinkSync(fs.readlinkSync(src), dst);
      } else {
        fs.copyFileSync(src, dst);
      }
    } else if (stat.isDirectory()) {
      copyDirRecursive(src, dst, resolveSymlinks);
    } else if (stat.isSymbolicLink()) {
      // Symlink pointing to a directory
      fs.mkdirSync(path.dirname(dst), { recursive: true });
      if (resolveSymlinks) {
        const real = fs.realpathSync(src);
        copyDirRecursive(real, dst, false);
      } else {
        fs.symlinkSync(fs.readlinkSync(src), dst);
      }
    }
  }

  return staging;
}

function copyDirRecursive(src: string, dst: string, resolveSymlinks: boolean): void {
  fs.mkdirSync(dst, { recursive: true });

  for (const entry of fs.readdirSync(src)) {
    const srcPath = path.join(src, entry);
    const dstPath = path.join(dst, entry);
    const stat = fs.lstatSync(srcPath);

    if (stat.isSymbolicLink()) {
      if (resolveSymlinks) {
        let real: string;
        try {
          real = fs.realpathSync(srcPath);
        } catch {
          // Skip broken symlinks
          continue;
        }
        const realStat = fs.statSync(real);
        if (realStat.isDirectory()) {
          copyDirRecursive(real, dstPath, false);
        } else if (realStat.isFile()) {
          fs.copyFileSync(real, dstPath);
        }
      } else {
        fs.symlinkSync(fs.readlinkSync(srcPath), dstPath);
      }
    } else if (stat.isDirectory()) {
      copyDirRecursive(srcPath, dstPath, resolveSymlinks);
    } else if (stat.isFile()) {
      fs.copyFileSync(srcPath, dstPath);
    }
  }
}

export function collectFiles(baseDir: string, syncPaths: string[]): Record<string, string> {
  const files: Record<string, string> = {};

  for (const sp of syncPaths) {
    const src = path.join(baseDir, sp);
    if (!fs.existsSync(src)) continue;

    const stat = fs.statSync(src);
    if (stat.isFile()) {
      try {
        files[sp] = fs.readFileSync(src, "utf-8");
      } catch {
        files[sp] = "<binary>";
      }
    } else if (stat.isDirectory()) {
      collectDir(src, baseDir, files);
    }
  }

  return files;
}

function collectDir(dir: string, baseDir: string, files: Record<string, string>): void {
  for (const entry of fs.readdirSync(dir).sort()) {
    const full = path.join(dir, entry);
    const stat = fs.statSync(full);
    if (stat.isFile()) {
      const rel = path.relative(baseDir, full);
      try {
        files[rel] = fs.readFileSync(full, "utf-8");
      } catch {
        files[rel] = "<binary>";
      }
    } else if (stat.isDirectory()) {
      collectDir(full, baseDir, files);
    }
  }
}

export function diffTrees(
  localFiles: Record<string, string>,
  remoteFiles: Record<string, string>,
): string[] {
  const allPaths = [...new Set([...Object.keys(localFiles), ...Object.keys(remoteFiles)])].sort();
  const diffs: string[] = [];

  for (const p of allPaths) {
    const localContent = localFiles[p] ?? "";
    const remoteContent = remoteFiles[p] ?? "";

    if (localContent === remoteContent) continue;

    if (!(p in remoteFiles)) {
      diffs.push(`  + ${p} (local only)`);
      continue;
    }

    if (!(p in localFiles)) {
      diffs.push(`  - ${p} (remote only)`);
      continue;
    }

    const patch = createTwoFilesPatch(
      `remote/${p}`,
      `local/${p}`,
      remoteContent,
      localContent,
    );
    if (patch) {
      diffs.push(patch);
    }
  }

  return diffs;
}

export function gitStatusForPaths(
  repoDir: string,
  syncPaths: string[],
  gitEnv?: Record<string, string>,
): { modified: string[]; ahead: number; behind: number } {
  const env = { ...process.env, ...(gitEnv ?? {}) };

  // Get status for sync paths
  const statusResult = spawnSync(
    "git",
    ["status", "--porcelain", "--", ...syncPaths],
    { cwd: repoDir, env, encoding: "utf-8" },
  );
  const modified: string[] = [];
  for (const line of (statusResult.stdout ?? "").trim().split("\n")) {
    if (line.trim()) {
      modified.push(line.trim());
    }
  }

  // Get ahead/behind
  let ahead = 0;
  let behind = 0;
  const revResult = spawnSync(
    "git",
    ["rev-list", "--left-right", "--count", "HEAD...@{upstream}"],
    { cwd: repoDir, env, encoding: "utf-8" },
  );
  if (revResult.status === 0 && revResult.stdout?.trim()) {
    const parts = revResult.stdout.trim().split(/\s+/);
    if (parts.length === 2) {
      ahead = parseInt(parts[0], 10);
      behind = parseInt(parts[1], 10);
    }
  }

  return { modified, ahead, behind };
}

export function gitPush(
  repoDir: string,
  syncPaths: string[],
  message: string,
  gitEnv?: Record<string, string>,
  dryRun = false,
): { commit_hash: string | null; files_staged: string[] } {
  const env = { ...process.env, ...(gitEnv ?? {}) };

  // Stage files
  spawnSync("git", ["add", "--", ...syncPaths], { cwd: repoDir, env });

  // Check if there's anything to commit
  const diffResult = spawnSync(
    "git",
    ["diff", "--cached", "--name-only"],
    { cwd: repoDir, env, encoding: "utf-8" },
  );
  const staged = (diffResult.stdout ?? "").trim().split("\n").filter(Boolean);

  if (staged.length === 0) {
    return { commit_hash: null, files_staged: [] };
  }

  if (dryRun) {
    // Unstage
    spawnSync("git", ["reset", "HEAD", "--", ...syncPaths], { cwd: repoDir, env });
    return { commit_hash: "(dry-run)", files_staged: staged };
  }

  // Commit
  spawnSync("git", ["commit", "-m", message], { cwd: repoDir, env });

  // Get commit hash
  const hashResult = spawnSync(
    "git",
    ["rev-parse", "--short", "HEAD"],
    { cwd: repoDir, env, encoding: "utf-8" },
  );
  const commitHash = (hashResult.stdout ?? "").trim();

  // Push
  spawnSync("git", ["push"], { cwd: repoDir, env });

  return { commit_hash: commitHash, files_staged: staged };
}

export function gitPull(
  repoDir: string,
  gitEnv?: Record<string, string>,
  dryRun = false,
): string {
  const env = { ...process.env, ...(gitEnv ?? {}) };

  if (dryRun) {
    const result = spawnSync(
      "git",
      ["fetch", "--dry-run"],
      { cwd: repoDir, env, encoding: "utf-8" },
    );
    return (result.stderr ?? "").trim() || "Already up to date.";
  }

  const result = spawnSync(
    "git",
    ["pull"],
    { cwd: repoDir, env, encoding: "utf-8" },
  );
  if (result.status !== 0) {
    throw new Error((result.stderr ?? "").trim() || "git pull failed");
  }
  return (result.stdout ?? "").trim() || "Already up to date.";
}
